### PR TITLE
feat: add ability to mock WebSocket connections in Cypress #949

### LINKED ISSUE #949: Add ability to mock WebSocket connections in Cypress
# **IMPORTANT: NO BOUNTIES WILL BE PAID FROM 2/5 - 2/14 - IF YOU COMPLETE THIS BOUNTY DURING THIS TIMEFRAME WE WILL REMIT PAYMENT ON 2/14**

# Background 

[Cypress](https://docs.cypress.io/guides/overview/why-cypress) supports [intercepting and mocking REST communications](https://docs.cypress.io/guides/guides/network-requests) out of the box. It does not, however, support WebSocket connections in the same way.

We use WebSockets to provide data to the web Dashboard, notably transaction data relating to the connected wallet. We are not currently able to test the rendering of transactions due to this Cypress limitation.

# Other considerations

A number of libraries exist that may be of use, e.g.:

- [cypress-websocket-testing](https://github.com/lensesio/cypress-websocket-testing)
- [mock-socket](https://github.com/thoov/mock-socket)
- ...

An example response from a WebSocket server relating to transactions is available in `cypress/factories/ethereum/transactions.ts`

We have several intercepted REST endpoints [here](https://github.com/shapeshift/web/blob/08f03825e01401033b94da0903713c8c39ec9229/cypress/support/commands.ts#L112) that could be used for inspiration - we probably want to have a Cypress WebSocket command implemented similarly.

# Acceptance Criteria

- Add the ability to mock WebSocket connections for Cypress tests
- Implement an example of a mock WebSocket server for transactions, with an appropriate test to match (e.g. testing that the mocked transactions render on the Dashboard correctly in `TransactionRow` when used in `RecentTransactions`, `TxHistory`, etc). This test will likely sit in `cypress/integration/dashboard_spec.ts`, though could be broken out if the tests are substantial enough.
- The implementation should be extensible and show us the best practice for interacting with WebSockets from Cypress moving forward

# Bounty Documentation 
- Please check out [our docs](https://shapeshiftdao.gitbook.io/getting-started/how-to-contribute/bounties) for how to get in touch with us or ask questions.

### PR DESCRIPTION
**Pull request by 0xApotheosis**
Recreated from @visitskyworld's [PR](https://github.com/shapeshift/web/pull/1013) to resolve merge conflicts and rebasing issues.

## Description

- Add the ability to mock WebSocket connections for Cypress tests
- Add `mockWebSocketRequest` and `mockAllWebSocketRequests` commands
- Implement an example of a mock WebSocket server for transactions - WIP

## Notice

Before submitting a pull request, please make sure you have answered the following:

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

 `closes #949`

## Testing
run
`yarn test:cypress`

### CODE REVIEW

File: cypress/support/commands.ts [hunk 1]
```diff
@@ -17,11 +21,24 @@ const seed = Cypress.env('testSeed')
 const publicKey = Cypress.env('testPublicKey')
 const ethereumApi = Cypress.env('REACT_APP_UNCHAINED_ETHEREUM_HTTP_URL')
 const bitcoinApi = Cypress.env('REACT_APP_UNCHAINED_BITCOIN_HTTP_URL')
+const wsUrl = Cypress.env('wsUrl')
 const _0xApi = Cypress.env('0xApi')
 const coinGeckoApi = Cypress.env('coinGeckoApi')
 const foxContract = Cypress.env('foxContract')
 
 const walletDb = getWalletDbInstance()
+const ethTransaction = makeEthTxHistory()
```
Comment (0xApotheosis) [inline cypress/support/commands.ts:30]: As we are returning multiple ETH transactions, should this be plural?

`ethTransactions`

File: tsconfig.json [hunk 2]
```diff
@@ -9,7 +9,7 @@
     "strict": true,
     "forceConsistentCasingInFileNames": true,
     "noFallthroughCasesInSwitch": true,
-    "module": "esnext",
+    "module": "commonjs",
```
Comment (0xApotheosis) [inline tsconfig.json:12]: What is this for? Is it necessary?

**Review by 0xApotheosis — COMMENTED** [review]
@visitskyworld we're getting there - my specific comments are all minor, though crucially I was not able to get the new websocket tests working.

I can see the WS connection and messages on `localhost`, but Cypress doesn't seem to intercept / recognise them.

Is there a config I'm missing somewhere?

Note - I can't "Request changes", as this PR is created by me.

**Review by 0xApotheosis — COMMENTED** [review]
@visitskyworld at the moment, the WebSocket server in `cypress/server/index.ts` is not extensible - our `subscribe` logic has no `data` param to specify the `topic` we want to subscribe to, or the related `addresses`.

I.e. the server is hardcoded to always return what usually comes back from a subscription to the `txs` topic.

Are we able to make the mock WebSocket server more generic?

**Review by 0xApotheosis — COMMENTED** [review]
As I understand it, the instance of `web` Cypress is talking to needs to point to the local, mocked, WebSocket server (`ws://localhost:8080`).

We can handle this by updating the CI configuration, though my concern is that this causes _all_ WebSocket requests to be intercepted.

Can we implement a pass-through mechanism, where if we haven't mocked a subscription topic, we'll send that subscription request through to the actual server?

This would require the more generic WebSocket server logic noted above, and will prevent Cypress from breaking when we add new WebSocket topics that are not explicitly mocked.

**Comment by visitskyworld** [comment]
@0xApotheosis , I have made changes you requested.
Regarding intercepting websocket, I tried to research possible solutions but there were no ways to intercept it.
The only solution is to configure cypress environment, I think.

**Comment by visitskyworld** [comment]
Also, I have no permission to push `cypress-websocket` branch so I made a new PR here.
https://github.com/shapeshift/web/pull/1099

**Comment by 0xApotheosis** [comment]
> @0xApotheosis , I have made changes you requested. Regarding intercepting websocket, I tried to research possible solutions but there were no ways to intercept it. The only solution is to configure cypress environment, I think.

Thanks @visitskyworld - we'll take what we have so far and move the final piece of the implementation in-house.

Appreciate your contributions on this one.

**Pull request closed without merging**